python/src/equistore/operations/join.py

```diff
@@ -80,155 +43,28 @@ def join(tensor_maps: List[TensorMap], axis: str):
-
-    # Label names are unique: We can do an equal check only checking the lengths.
-    names_are_same = np.all(
-        len(unique_names) == np.array([len(names) for names in names_list])
-    )
-
-    if names_are_same:
-        unique_values = np.unique(np.vstack(values_list), axis=0)
-        values_are_unique = len(unique_values) == len(np.vstack(values_list))
-        if values_are_unique:
-            new_names = names_list[0]
-            new_values = np.vstack(values_list)
-        else:
-            new_names = ["tensor"] + list(names_list[0])
-            new_values = np.hstack(
-                [tensor_values.reshape(-1, 1), np.vstack(values_list)]
-            )
-    else:
-        new_names = ["tensor", "property"]
-        property_values = np.hstack([np.arange(len(values)) for values in values_list])
-        new_values = np.vstack([tensor_values, property_values]).T
-
-    return Labels(names=new_names, values=new_values)
-
-
-def _join_blocks_along_properties(blocks: List[TensorBlock]) -> TensorBlock:
-    """Join a sequence of :py:class:`TensorBlock` along their properties."""
-    first_block = blocks[0]
-
-    fname = "_join_blocks_along_properties"
-    for block in blocks[1:]:
-        _check_blocks(first_block, block, ["components", "samples"], fname)
-        _check_same_gradients(first_block, block, ["components", "samples"], fname)
-
-    properties = _join_labels([block.properties for block in blocks])
-    result_block = TensorBlock(
-        values=_dispatch.hstack([block.values for block in blocks]),
-        samples=first_block.samples,
-        components=first_block.components,
-        properties=properties,
-    )
-
-    for parameter, first_gradient in first_block.gradients():
-        # Different blocks might contain different gradient samples
-        gradient_sample_values = np.unique(
-            np.hstack([block.gradient(parameter).samples for block in blocks])
-        ).tolist()
-        gradient_samples = Labels(
-            names=first_gradient.samples.names, values=np.array(gradient_sample_values)
-        )
-
-        gradient_data = np.zeros(
-            (len(gradient_samples),)
-            + first_gradient.data.shape[1:-1]
-            + (len(properties),)
-        )
-
-        properties_start = 0
-        for block in blocks:
-            gradient = block.gradient(parameter)
 
-            # find the correct sample position to put the data
-            for i_grad, sample_grad in enumerate(gradient.samples):
-                i_sample = gradient_samples.position(sample_grad)
-                gradient_data[
-                    i_sample,
-                    ...,
-                    properties_start : properties_start + len(gradient.properties),
-                ] = gradient.data[i_grad]
+    for i, tensor in enumerate(tensor_maps):
+        keys_values += [value + (i,) for value in tensor.keys.tolist()]
+        blocks += [block.copy() for _, block in tensor]
```

Comment:
After thinking a bit more about it, I think this is where we should remove the copy. Instead of using block.copy(), recreating the block with 

```
TensorBlock(
    block.values,
    block.samples,
    block.components,
    block.properties,
)
```

should only increment the reference count of everything, not making any copy, and just create a new block sharing data. Then we have one unique copy below, which is required anyway.

Reply:
If this will just increase the count, this should work here!

Reply:
Yes, I did a test for a spherical expansion and an atomic composition. This worked. Also the length 0 gradients of the atomic composition are densified.